src/main/java/org/springframework/data/r2dbc/function/StatementFactory.java

```diff
@@ -0,0 +1,102 @@
+ * limitations under the License.
+ */
+package org.springframework.data.r2dbc.function;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+import org.springframework.data.r2dbc.dialect.Dialect;
+import org.springframework.data.r2dbc.domain.SettableValue;
+import org.springframework.data.relational.core.sql.Delete;
+import org.springframework.data.relational.core.sql.Insert;
+import org.springframework.data.relational.core.sql.Select;
+import org.springframework.data.relational.core.sql.Update;
+
+/**
+ * Interface declaring statement methods that are commonly used for {@code SELECT/INSERT/UPDATE/DELETE} operations.
+ * These methods consider {@link Dialect} specifics and accept bind parameters with values.
+ *
+ * @author Mark Paluch
+ * @see PreparedOperation
+ */
+public interface StatementFactory {
+
+	/**
+	 * Creates a {@link Select} statement.
+	 *
+	 * @param tableName must not be {@literal null} or empty.
+	 * @param columnNames the columns to project, must not be {@literal null} or empty.
+	 * @param binderConsumer customizer for bindings. Supports only
+	 *          {@link StatementBinderBuilder#filterBy(String, SettableValue)} bindings.
+	 * @return the {@link PreparedOperation} to select the given columns.
+	 */
+	PreparedOperation<Select> select(String tableName, Collection<String> columnNames,
+			Consumer<StatementBinderBuilder> binderConsumer);
+
+	/**
+	 * Creates a {@link Insert} statement.
+	 *
+	 * @param tableName must not be {@literal null} or empty.
+	 * @param generatedKeysNames must not be {@literal null}.
+	 * @param binderConsumer customizer for bindings. Supports only
+	 *          {@link StatementBinderBuilder#bind(String, SettableValue)} bindings.
+	 * @return the {@link PreparedOperation} to update values in {@code tableName} assigning bound values.
+	 */
+	PreparedOperation<Insert> insert(String tableName, Collection<String> generatedKeysNames,
+			Consumer<StatementBinderBuilder> binderConsumer);
+
+	/**
+	 * Creates a {@link Update} statement.
+	 *
+	 * @param tableName must not be {@literal null} or empty.
+	 * @param binderConsumer customizer for bindings.
+	 * @return the {@link PreparedOperation} to update values in {@code tableName} assigning bound values.
+	 */
+	PreparedOperation<Update> update(String tableName, Consumer<StatementBinderBuilder> binderConsumer);
+
+	/**
+	 * Creates a {@link Delete} statement.
+	 *
+	 * @param tableName must not be {@literal null} or empty.
+	 * @param binderConsumer customizer for bindings. Supports only
+	 *          {@link StatementBinderBuilder#filterBy(String, SettableValue)} bindings.
+	 * @return the {@link PreparedOperation} to delete rows from {@code tableName}.
+	 */
+	PreparedOperation<Delete> delete(String tableName, Consumer<StatementBinderBuilder> binderConsumer);
+
+	/**
+	 * Binder to specify parameter bindings by name. Bindings match to equals comparisons.
+	 */
+	interface StatementBinderBuilder {
```

Comment:
I'm not a fan of that name. I made me look for a `build` method returning a `StatementBinder` and didn't help me otherwise. 

How about a `SomethingCustomizer` or `CustomizableSomething`. `Something` might actually be `Query`, `Statement` or `Operation`. Again a point in the case that we need to clarify these terms.